src/lib.rs

```diff
@@ -403,24 +399,18 @@ impl<D: Domain> NoteEncryption<D> {
     }
 
     /// Generates `encCiphertext` for this note.
-    pub fn encrypt_note_plaintext(&self) -> [u8; ENC_CIPHERTEXT_SIZE] {
+    pub fn encrypt_note_plaintext(&self) -> D::NoteCiphertextBytes {
         let pk_d = D::get_pk_d(&self.note);
         let shared_secret = D::ka_agree_enc(&self.esk, &pk_d);
         let key = D::kdf(shared_secret, &D::epk_bytes(&self.epk));
-        let input = D::note_plaintext_bytes(&self.note, &self.memo);
+        let mut input = D::note_plaintext_bytes(&self.note, &self.memo);
+
+        let output = input.as_mut();
 
-        let mut output = [0u8; ENC_CIPHERTEXT_SIZE];
-        output[..NOTE_PLAINTEXT_SIZE].copy_from_slice(&input.0);
         let tag = ChaCha20Poly1305::new(key.as_ref().into())
-            .encrypt_in_place_detached(
-                [0u8; 12][..].into(),
-                &[],
-                &mut output[..NOTE_PLAINTEXT_SIZE],
-            )
+            .encrypt_in_place_detached([0u8; 12][..].into(), &[], output)
             .unwrap();
-        output[NOTE_PLAINTEXT_SIZE..].copy_from_slice(&tag);
-
-        output
+        D::NoteCiphertextBytes::from((output, tag.as_ref()))
```

Comment:
Once the `From`s are replaced, this becomes
```suggestion
        D::parse_note_ciphertext_bytes(output, tag).expect("output is correct length")
```
the unwrap here being fine because we know that `output` is derived from `D::NotePlaintextBytes`, which we can assume is the correct size relative to `D::NoteCiphertextBytes`.

Non-blocking: I have ideas on how to make this infallible in a type-safe way, but I'll look at that refactor after this PR lands.

Reply:
Done (added and used `parse..` method).